### ISSUE TITLE
Keyboard in block sectors?

### ISSUE THREAD

**Issue by OneBadNinja** [OptiKey/OptiKey#327]
Hey there Julius, 

A family member of mine is an ALS patient and she's having a hard time coping with the keyboard (elderly lady) to communicate properly.

I was thinking it would be much easier for people in her case to have a simpler interface - keyboard blocks to communicate with (perhaps a new communication mode from the menu).

For instance there are initially 6 blocks on the screen, each containing a group of letters. The user picks the block that contains their letter, and then that block is further dissected recursively until there are large blocks each containing a single letter (emphasis on large blocks as she struggles with focusing on the smaller blocks)

There is a video on youtube here that contains a similar functionality, whereas you slowly zoom into the desired letter:

https://www.youtube.com/watch?v=LtSCemhpIWk

Only simple letter blocks would be needed. She finds more detailed menus very confusing, and I imagine could potentially be the same for some others her age.

Would this be feasible to make? Or does the software already have this feature and I've missed it?

Thank you so much,
Joey

**Comment by zelmon64** [comment]
Hi Joey,

To my knowledge this hasn't been done before but should be relatively simple to implement. If you specify the arrangement of letters you want I should be able to do it.

As an already existing alternative there is [this Yes/No build](https://github.com/OptiKey/OptiKey/issues/322#issuecomment-301759584) which is usable but not quite finished.

**Comment by JuliusSweetland** [comment]
@OneBadNinja Hi Joey. No this method of input does not currently exist. If @Zelmon64 is kind enough to add it then that would be great. Otherwise there is work underway to allow users to define their own keyboards. It would be possible to create and share the keyboard you are describing using this new method, but it is not ready for release yet. The work is being done by someone else so I cannot give a timeline I'm afraid.

**Comment by OneBadNinja** [comment]
Hi Julius and Zelmon, thank you for both of your replies. I understand the situation now. Also apologies for getting your name wrong earlier Julius. 

Zelmon I would be sincerely grateful if you would be able to help out. Thank you so much for volunteering! To offer some clarity, I have just designed a small mockup of this simple interface, and I have attached the images below. I hope this isn't too troublesome

## The Main Grid

User selects a block containing letters/symbols and is taken to the next pane.

![pane_1](https://user-images.githubusercontent.com/20068286/27964943-13471db4-634b-11e7-8fa9-367b4fc109af.png)

## Secondary Pane

User now further selects a smaller subsection of the initial letter group. 

![pane_2](https://user-images.githubusercontent.com/20068286/27964946-17c54dc0-634b-11e7-8dda-e3c3587a02ce.png)

## Final Pane

Each letter now has its own large block. The user finally picks one and it's written to the board. Previous button takes the user back to the second image, and main grid takes the user to the first image.

![pane_3](https://user-images.githubusercontent.com/20068286/27964949-1bcd733e-634b-11e7-8eea-9f6af0838fd2.png)

## Aux

If the user were to select the bottom left block in the first image, this secondary pane would appear, where the top row has 3 columns now due to 3 rows of letters in the initial block.

![pane_4](https://user-images.githubusercontent.com/20068286/27964953-21a4c0b4-634b-11e7-82cd-737d4b688dad.png)

I hope this is helpful, let me know if you need anything!

**Comment by JuliusSweetland** [comment]
@OneBadNinja No problem about the name. One question - how does the user speak? There is a dedicated 'Speak' key in OptiKey. One possibility would be to add a clear key over the scratchpad itself, but this may be confusing or cause the midas touch problem as the user checks what they have typed.

**Comment by OneBadNinja** [comment]
We're from Georgia, and hence the setup we're after at the moment is typing Georgian letters using the latin alphabet, and the nurses reading the message on the screen. The speak key is a bit redundant with pronouncing Georgian words.

However I understand that the more general this is made the better it would be for other users who may want the speech function, so perhaps something like this? The main intent of this layout would be for simple speech in any case, so not much use of symbols or numbers, hence the smaller size. Also I've added caps though I wouldn't see it as much use in our case.

## Generalised

![speak](https://user-images.githubusercontent.com/20068286/27973940-e3c4314a-636c-11e7-8078-ccb8d6b4c3a3.png)

**Comment by JuliusSweetland** [comment]
I think that looks really good. Why would you not want to speak Georgian?

**Comment by OneBadNinja** [comment]
Glad to hear that. We wouldn't use the speak function as Georgian words aren't accurately pronounced. In terms of the alphabet, I supposed that the latin alphabet would have been an easier implementation for this mode at this stage, let alone more beneficial to other users too. It should be sufficient at the moment.

I am currently a part of the Transifex team for localising the main app into Georgian, and I could certainly help with localising this too should the situation require. We would be fine trying it out with English first, not a problem

**Comment by JuliusSweetland** [comment]
Are they not pronounced correctly because you are using an English voice,
or another reason?

**Comment by zelmon64** [comment]
Here's what I've done so far. It's a little different to your template. I though it best to standardise with three main alpha buttons per layer.

To enable this keyboard select `Blocky letters` as the stastup keyboard in the management console.

![image](https://user-images.githubusercontent.com/15001389/27988403-fa89c672-6418-11e7-8ab9-c19eb69736aa.png)
https://drive.google.com/open?id=0B7SYYOqNW4FSckpWdUtBQXZoOWs

**Comment by JuliusSweetland** [comment]
@zelmon64 Nice work. If you want to break apart the Output control you could extract just what you want, i.e. CLEAR, SCRATCHPAD, SPEAK and SLEEP;

```
        <controls:Key Grid.Row="1" Grid.Column="1"
                      SymbolGeometry="{StaticResource ClearIcon}" 
                      Text="{x:Static resx:Resources.CLEAR}"
                      SharedSizeGroup="KeyWithSymbol"
                      Value="{x:Static models:KeyValues.ClearScratchpadKey}"/>
        <controls:Scratchpad Grid.Row="1" Grid.Column="2"
                             Text="{Binding DataContext.KeyboardOutputService.Text, RelativeSource={RelativeSource AncestorType=controls:KeyboardHost}, Mode=OneWay}" />
        <controls:Key Grid.Row="1" Grid.Column="3"
                      SymbolGeometry="{StaticResource SpeakIcon}" 
                      Text="{x:Static resx:Resources.SPEAK}"
                      SharedSizeGroup="KeyWithSymbol"
                      Value="{x:Static models:KeyValues.SpeakKey}"/>
```

**Comment by zelmon64** [comment]
@JuliusSweetland Thanks.

Thanks for that snippet. I like having the prediction and speak options but I was forgetting that there isn't a Georgian dictionary yet so I'll make another build without it.

**Comment by JuliusSweetland** [comment]
@zelmon64 Thanks. The other issue with the smaller keys is that I presume the users who will want the larger keys will struggle with the small Output key sizes.

**Comment by zelmon64** [comment]
@JuliusSweetland I agree. Is there a way to increase the scratchpad font? It's not using the full height.

![image](https://user-images.githubusercontent.com/15001389/27988583-e29e7284-641c-11e7-966c-6be8d99a4f1e.png)

**Comment by JuliusSweetland** [comment]
@zelmon64 Yes, ScratchpadNumberOfLines in the Visual tab of the Management Studio. Set it to 1.

You can, if desired, normalise the icon sizes by specifying the WidthSpan and/or HeightSpan on the keys which span more than 1 of the smallest key size. From the above screenshot you could set the WidthSpan=2 and HeightSpan=2 on both the back keys. You can also set WidthSpan=2 on the space key. Give it a try and see what you think. It will have its most clear impact when you hover over keys as the key text will also be normalised.

**Comment by zelmon64** [comment]
@JuliusSweetland the space key already had `WidthSpan=2` but the corner icons are nicer now thanks. Here's an image of all three layers. Anything else you'd suggest?

![image](https://user-images.githubusercontent.com/15001389/27988661-188c3f78-641f-11e7-8249-375f54ac6853.png)
![image](https://user-images.githubusercontent.com/15001389/27988664-27391028-641f-11e7-82e0-8314ddc7686d.png)
![image](https://user-images.githubusercontent.com/15001389/27988666-381a0b68-641f-11e7-9af7-292fa4b5aae5.png)

**Comment by OneBadNinja** [comment]
Ah guys this is incredible.. thank you so much Zelman! This is very exciting!, I will be passing by tomorrow and we'll give this a shot!

A tiny observation in the earlier file: the caps lock key doesn't seem to be working. I figured it would be easier to have the option to keep it in capitals just so it isn't confusing in the start, if you could fix that that would be awesome!

Julius, I'm guessing the pronunciation must be wrong because it's reading the Georgian words written with the latin alphabet, and hence it's not reading it with Georgian phonemes. Potentially once I localise in Georgian I think it should be able to read the symbols if the default Anna voice is able to (I just tested this on eSpeak, but I assume its the same voice library?).

Just saw the latest version: that looks very clean! However, the sleep key is missing now, would it be possible to add that back in? Spatially, it could replace caps if the scratchboard could persist as capitalised.

**Comment by JuliusSweetland** [comment]
@zelmon64 It looks really good. Couple of suggestions though...

I really like the layout of this: 
![speak](https://user-images.githubusercontent.com/20068286/27973940-e3c4314a-636c-11e7-8078-ccb8d6b4c3a3.png)

1.Would it be possible to more closely match it (maybe with the speak and backspace keys swapped)?
2.I don't think the bottom left key should change depending on which keyboard level you are at. At lower levels (not the main keyboard) I like the back to main key (maybe use the ABC icon called 'AlphaIcon') and the back key.
3.Would it be possible to create the numeric and symbols keyboard in the same layout? Perhaps the 4 top level keys could be:
123 | 456 |7890 | !?.
I don't think too many symbols (in English anyway) would be needed so we could have just 2 levels. What do you think?

**Comment by zelmon64** [comment]
@OneBadNinja you're very welcome.

How's this? The only punctuation is a question mark.

![image](https://user-images.githubusercontent.com/15001389/27988810-e513ed44-6423-11e7-8291-6243983af1c3.png)

https://drive.google.com/open?id=0B7SYYOqNW4FSbUVULWZDaHVEM2s

@JuliusSweetland I like your suggestions. I'll have a go

**Comment by OneBadNinja** [comment]
Zelmon this is perfect! Exactly what we need!! Thank you so, SO much!!! As simple as it can be, no need for other punctuation marks! I can't wait to try this out tomorrow! Let you both know how it goes!

**Comment by zelmon64** [comment]
@JuliusSweetland this is the layout I have so far. What do you think?

![image](https://user-images.githubusercontent.com/15001389/27989659-805859e6-6435-11e7-8973-2bced951ecd4.png)
![image](https://user-images.githubusercontent.com/15001389/27989663-96e4ddf6-6435-11e7-9bea-068318f24190.png)
![image](https://user-images.githubusercontent.com/15001389/27989667-a736e83e-6435-11e7-9775-8786e33af4eb.png)

https://drive.google.com/open?id=0B7SYYOqNW4FSUnFtSXZNVXhkZWs

**Comment by JuliusSweetland** [comment]
Looking good! What do you think about swapping the 4th letter/number key in
the 2nd level keyboards with the alpha ABC key? That way every
letter/number key is the same size.

Excellent work.

**Comment by zelmon64** [comment]
@JuliusSweetland that's a good idea. I've offset the cluster selection texts to indicate the new position as well. Here are all the layers:

![image](https://user-images.githubusercontent.com/15001389/27992341-5badd422-648a-11e7-9f59-6d5a6262a721.png)
![image](https://user-images.githubusercontent.com/15001389/27992369-eeabe5b6-648a-11e7-838e-eb5547df4916.png)
![image](https://user-images.githubusercontent.com/15001389/27992372-06a895c4-648b-11e7-965c-5bf21e4d12a0.png)
![image](https://user-images.githubusercontent.com/15001389/27992337-4ab28852-648a-11e7-8631-b24341f8b605.png)
![image](https://user-images.githubusercontent.com/15001389/27992324-e1110af4-6489-11e7-8dc4-185723c85e67.png)

**Comment by JuliusSweetland** [comment]
That looks excellent. 

@OneBadNinja can you test the very latest version and we can merge it and create an official release if you are happy. In the same release will be an even simpler YES/NO keyboard for people who need the most basic experience. Together they make a really valuable release so thank you everyone.

**Comment by JuliusSweetland** [comment]
@zelmon64 One more idea. How about a new setting EnableQuitKeys (defaulted to false), which if disabled means all quit keys can be disabled, or even removed. This would allow the bottom right Sleep key to span 2 keys and make the keyboard more balanced. I could then use this setting to do the same on other keyboards (i.e. the ConversationAlpha keyboards and the new ConversationConfirm keyboard). I think there's a good use case to say that the user shouldn't be able to accidentally quit the application as this may not be something that will ever be desired except when a carer quits (which can be done by right clicking on OptiKey and selecting QUIT from the context menu.

What do you think?

**Comment by zelmon64** [comment]
@JuliusSweetland I like the idea of EnableQuitKeys.

**Comment by OneBadNinja** [comment]
Julius and Zelmon you have our biggest thanks! We have just run the test and for the first time she's written not one, but two words: TINA and TOMA - her daughter and great grandson's names! She looks exhilarated! It has been years since she has been able to communicate independently and now she will slowly begin again, all thanks to you!

I have tested the latest version and it's great! Once you type in a number however it goes back to the main alphabetical menu - could this possibly go back to the main numbers panel instead? It's likely a user would want to type in a date or any number above one digit. Also agree with Julius about the sleep key extending over the quit key!

Another small observation: the caps lock effect seems to turn off after using the spacebar, could you please fix this for both versions to keep the uppercase? I say both because we'll slowly introduce her to the numbers menu as well, for now we'll focus on letters

I am very happy to hear you're planning to merge this into the latest release, i'm sure many will find this useful! Thank you both once again!

**Comment by zelmon64** [comment]
@OneBadNinja I'm so glad it was such a success! 

I've made the numbers return to the main numbers panel but the punctuation return to the main alphabetical menu. The keys should now be completely unaffected by the shift key state.

![image](https://user-images.githubusercontent.com/15001389/27994992-4ce7d4e4-64bf-11e7-9d1c-ec53523b360c.png)

https://drive.google.com/open?id=0B7SYYOqNW4FScXVEU2d2cjhMUW8

**Comment by JuliusSweetland** [comment]
@OneBadNinja is it possible to trigger the sleep key to span two keys of the quit key is disabled? I would do this by having two sleep keys - 1 spanning two widths, and toggle the visibility?

**Comment by zelmon64** [comment]
@JuliusSweetland I haven't figured out how to wrap differently sized buttons in triggers. At the moment the sizes are defined at the start with `ContentControl`. Do you have an example I could copy?

**Comment by JuliusSweetland** [comment]
If you define two keys - one with column span = 1 and the other with column
span = 2 then set a visibility trigger on each key with opposite logic,
i.e. The column span = 1 is visible if the quit key is visible, and the
other key is visible if not. I'm not at my laptop so I can't point you
towards a specific example, but there are visibility triggers all over the
place. Look for a Boolean to visibility value converter to find one. Let me
know if you can't and I'll dig one out for you.

**Comment by zelmon64** [comment]
@JuliusSweetland That worked thanks:
![image](https://user-images.githubusercontent.com/15001389/27996088-288ef43e-64d2-11e7-956c-b5c54b7ba1e0.png)

**Comment by JuliusSweetland** [comment]
@zelmon64 Great stuff. When you're happy submit a PR and I'll merge it in and make a release.

FYI I think I'm going to reverse the default value of the EnableQuitKeys setting (to true) so the change doesn't impact existing users.

**Comment by JuliusSweetland** [comment]
@OneBadNinja That's excellent feedback. I'm very happy to hear that. All credit goes to @zelmon64 for this one!

**Comment by zelmon64** [comment]
@JuliusSweetland the PR says it can't be automatically merge. Should I still submit the PR. I'm not sure what the procedure is.

I think the default value of the EnableQuitKeys is already true (or at least that's what I had intended it to be).

**Comment by JuliusSweetland** [comment]
You can either rebase from my master branch, (which will require you to fix
any merge conflicts) and then resubmit a PR, or just submit it and I'll
sort it out. I don't mind either way.

**Comment by zelmon64** [comment]
I'll try rebasing it and if I can't manage it I'll hand it over to you thanks.

**Comment by JuliusSweetland** [comment]
Sounds good, thank you.

I pushed some new changes yesterday so it'll be a conflict with something
in those changes. If you're not sure just ask.

**Comment by zelmon64** [comment]
@JuliusSweetland rebasing went really well so I have submitted the PR.

**Comment by zelmon64** [comment]
@OneBadNinja I have a new build which has more integration with the other keyboards if you'd like to try it: 
https://drive.google.com/open?id=0B7SYYOqNW4FSOTBNcjlYQnlub1U

To set it up like it use to be apply the following settings in the management console:

- `Visuals`
    - `Enable quit keys?` -> unchecked
    - `Enable Conversation Confirm  keyboard?` -> unchecked
    - `Conversation only mode:` -> checked
- `Words`
    - `Use simplefy keyboard layout:` ->  checked
    - `Force caps lock:` ->  checked

**Comment by OneBadNinja** [comment]
Zelmon this is phenomenal! Thank you so much for your work! Truly incredible and thank you both once again for your ceaseless commitment!

**Comment by JuliusSweetland** [comment]
This was released in v2.4.0
